File: packages/ingestion-market-data/src/provider.ts
import NodeCache from 'node-cache';
import type { MarketQuote, MarketDataProvider, MarketDataConfig } from './types.js';

export class MarketDataService implements MarketDataProvider {
  private alphaVantageKey?: string;
  private finnhubKey?: string;
  private polygonKey?: string;
  private cache?: NodeCache;
  private cacheMaxAgeHours: number;

  constructor(config: MarketDataConfig = {}) {
    this.alphaVantageKey = config.alphaVantageKey || process.env.ALPHA_VANTAGE_API_KEY;
    this.finnhubKey = config.finnhubKey || process.env.FINNHUB_API_KEY;
    this.polygonKey = config.polygonKey || process.env.POLYGON_API_KEY;
    this.cacheMaxAgeHours = config.cacheMaxAgeHours || 24;

    if (config.cacheEnabled !== false) {
      // Cache for 24 hours by default
      this.cache = new NodeCache({ stdTTL: this.cacheMaxAgeHours * 3600 });
    }
  }

  async getQuote(symbol: string): Promise<MarketQuote | null> {
    // Check cache first
    if (this.cache) {
      const cached = this.cache.get<MarketQuote>(symbol.toLowerCase());
      if (cached && !this.isCacheStale(symbol)) {
        return cached;
      }
    }

    // Try providers in order
    let quote: MarketQuote | null = null;

    // Try Alpha Vantage first
    if (this.alphaVantageKey) {
      try {
        quote = await this.fetchFromAlphaVantage(symbol);
      } catch (error) {
        console.warn(`Alpha Vantage failed for ${symbol}:`, error);
      }
    }

    // Fallback to Finnhub
    if (!quote && this.finnhubKey) {
      try {
        quote = await this.fetchFromFinnhub(symbol);
      } catch (error) {
        console.warn(`Finnhub failed for ${symbol}:`, error);
      }
    }

    // Final fallback to Polygon
    if (!quote && this.polygonKey) {
      try {
        quote = await this.fetchFromPolygon(symbol);
      } catch (error) {
        console.warn(`Polygon failed for ${symbol}:`, error);
      }
    }

    // Cache the result
    if (quote && this.cache) {
      this.cache.set(symbol.toLowerCase(), quote);
    }

    return quote;
  }

  async getQuotes(symbols: string[]): Promise<MarketQuote[]> {
    const quotes: MarketQuote[] = [];

    // Fetch in parallel for better performance
    const promises = symbols.map(symbol => this.getQuote(symbol));
    const results = await Promise.allSettled(promises);

    for (const result of results) {
      if (result.status === 'fulfilled' && result.value) {
        quotes.push(result.value);
      }
    }

    return quotes;
  }

  getCachedQuote(symbol: string): MarketQuote | null {
    if (!this.cache) return null;
    return this.cache.get<MarketQuote>(symbol.toLowerCase()) || null;
  }

  isCacheStale(symbol: string, maxAgeHours?: number): boolean {
    const maxAge = maxAgeHours || this.cacheMaxAgeHours;
    const cached = this.getCachedQuote(symbol);

    if (!cached) return true;

    const ageSeconds = (Date.now() / 1000) - cached.timestamp;
    return ageSeconds > (maxAge * 3600);
  }

  private async fetchFromAlphaVantage(symbol: string): Promise<MarketQuote> {
    const url = 'https://www.alphavantage.co/query';
    const params = new URLSearchParams({
      function: 'GLOBAL_QUOTE',
      symbol: symbol.toUpperCase(),
      apikey: this.alphaVantageKey!
    });

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000);

    try {
      const response = await fetch(`${url}?${params}`, {
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`Alpha Vantage API error: ${response.status}`);
      }

      const data: any = await response.json();

      if (!data['Global Quote']) {
        throw new Error('Invalid response from Alpha Vantage');
      }

      const quote = data['Global Quote'];
      return {
        symbol: symbol.toUpperCase(),
        price: parseFloat(quote['05. price'] || '0'),
        change: parseFloat(quote['09. change'] || '0'),
        changePercent: quote['10. change percent'] || '0%',
        volume: parseInt(quote['06. volume'] || '0'),
        lastUpdated: quote['07. latest trading day'] || '',
        source: 'alpha_vantage',
        timestamp: Math.floor(Date.now() / 1000)
      };
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async fetchFromFinnhub(symbol: string): Promise<MarketQuote> {
    const url = 'https://finnhub.io/api/v1/quote';
    const params = new URLSearchParams({
      symbol: symbol.toUpperCase(),
      token: this.finnhubKey!
    });

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000);

    try {
      const response = await fetch(`${url}?${params}`, {
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`Finnhub API error: ${response.status}`);
      }

      const data: any = await response.json();

      // Validate required fields
      if (typeof data.c !== 'number' || typeof data.d !== 'number') {
        throw new Error('Invalid response from Finnhub');
      }

      return {
        symbol: symbol.toUpperCase(),
        price: data.c,
        change: data.d,
        changePercent: `${data.dp?.toFixed(2) || 0}%`,
        high: data.h,
        low: data.l,
        open: data.o,
        previousClose: data.pc,
        lastUpdated: new Date(data.t * 1000).toISOString().split('T')[0],
        source: 'finnhub',
        timestamp: Math.floor(Date.now() / 1000)
      };
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async fetchFromPolygon(symbol: string): Promise<MarketQuote> {
    const url = `https://api.polygon.io/v2/aggs/ticker/${symbol.toUpperCase()}/prev`;
    const params = new URLSearchParams({
      apiKey: this.polygonKey!
    });

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000);

    try {
      const response = await fetch(`${url}?${params}`, {
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`Polygon API error: ${response.status}`);
      }

      const data: any = await response.json();

      if (data.status !== 'OK' || !data.results?.[0]) {
        throw new Error('Invalid response from Polygon');
      }

      const result = data.results[0];
      const change = result.c - result.o;
      const changePercent = result.o !== 0 ? ((change / result.o) * 100).toFixed(2) : '0';

      return {
        symbol: symbol.toUpperCase(),
        price: result.c,
        open: result.o,
        high: result.h,
        low: result.l,
        volume: result.v,
        change,
        changePercent: `${changePercent}%`,
        lastUpdated: new Date(result.t).toISOString().split('T')[0],
        source: 'polygon',
        timestamp: Math.floor(Date.now() / 1000)
      };
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

File: packages/ingestion-market-data-python/src/goblinos_ingestion_market_data/provider.py
"""
Market data provider implementation with caching and fallback logic.
"""

import time
import requests
from typing import Optional, List, Dict, Any
from .types import MarketQuote, MarketDataConfig


class MarketDataService:
    """Market data service with intelligent caching and provider fallback."""

    def __init__(self, config: Optional[MarketDataConfig] = None):
        self.config = config or {}
        self.alpha_vantage_key = self.config.get("alpha_vantage_key") or __import__(
            "os"
        ).getenv("ALPHA_VANTAGE_API_KEY")
        self.finnhub_key = self.config.get("finnhub_key") or __import__("os").getenv(
            "FINNHUB_API_KEY"
        )
        self.polygon_key = self.config.get("polygon_key") or __import__("os").getenv(
            "POLYGON_API_KEY"
        )
        self.cache_ttl = self.config.get("cache_ttl_seconds", 86400)  # 24 hours default
        self._cache: Dict[str, Dict[str, Any]] = {}

    def get_quote(self, symbol: str) -> Optional[MarketQuote]:
        """Get current quote for a symbol with caching."""
        # Check cache first
        cached = self._get_cached_quote(symbol)
        if cached:
            return cached

        # Try providers in order
        quote = None

        # Try Alpha Vantage first
        if self.alpha_vantage_key:
            try:
                quote = self._fetch_from_alpha_vantage(symbol)
            except Exception as e:
                print(f"Alpha Vantage failed for {symbol}: {e}")

        # Fallback to Finnhub
        if not quote and self.finnhub_key:
            try:
                quote = self._fetch_from_finnhub(symbol)
            except Exception as e:
                print(f"Finnhub failed for {symbol}: {e}")

        # Final fallback to Polygon
        if not quote and self.polygon_key:
            try:
                quote = self._fetch_from_polygon(symbol)
            except Exception as e:
                print(f"Polygon failed for {symbol}: {e}")

        # Cache the result
        if quote:
            self._cache_quote(symbol, quote)

        return quote

    def get_quotes(self, symbols: List[str]) -> List[MarketQuote]:
        """Get quotes for multiple symbols."""
        results = []
        for symbol in symbols:
            quote = self.get_quote(symbol)
            if quote:
                results.append(quote)
        return results

    def get_ohlc(
        self, symbol: str, timeframe: str, limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get OHLC data for a symbol and timeframe using Polygon."""
        if not self.polygon_key:
            return []

        # Map timeframe to Polygon parameters
        now = __import__("datetime").datetime.now()
        if timeframe == "1D":
            days_back = 1
            multiplier, timespan = 5, "minute"
        elif timeframe == "1W":
            days_back = 7
            multiplier, timespan = 1, "day"
        elif timeframe == "1M":
            days_back = 30
            multiplier, timespan = 1, "day"
        else:
            return []

        from_date = (now - __import__("datetime").timedelta(days=days_back)).strftime(
            "%Y-%m-%d"
        )
        to_date = now.strftime("%Y-%m-%d")

        url = f"https://api.polygon.io/v2/aggs/ticker/{symbol.upper()}/range/{multiplier}/{timespan}/{from_date}/{to_date}"
        params = {
            "adjusted": "true",
            "sort": "asc",
            "limit": min(limit, 50000),
            "apiKey": self.polygon_key,
        }

        try:
            response = requests.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()

            if data.get("status") == "OK" and data.get("results"):
                return [
                    {
                        "t": r["t"],  # timestamp
                        "o": r["o"],  # open
                        "h": r["h"],  # high
                        "l": r["l"],  # low
                        "c": r["c"],  # close
                        "v": r.get("v", 0),  # volume
                    }
                    for r in data["results"]
                ]
        except Exception as e:
            print(f"Polygon OHLC failed for {symbol}: {e}")

        return []

    def _get_cached_quote(self, symbol: str) -> Optional[MarketQuote]:
        """Get cached quote if still valid."""
        cached = self._cache.get(symbol)
        if cached and time.time() - cached["timestamp"] < self.cache_ttl:
            return cached["data"]
        return None

    def _cache_quote(self, symbol: str, quote: MarketQuote) -> None:
        """Cache a quote."""
        self._cache[symbol] = {"data": quote, "timestamp": time.time()}

    def _fetch_from_alpha_vantage(self, symbol: str) -> MarketQuote:
        """Fetch data from Alpha Vantage API."""
        url = "https://www.alphavantage.co/query"
        params = {
            "function": "GLOBAL_QUOTE",
            "symbol": symbol,
            "apikey": self.alpha_vantage_key,
        }

        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        if "Global Quote" not in data:
            raise ValueError("Invalid response from Alpha Vantage")

        quote = data["Global Quote"]
        return {
            "symbol": symbol,
            "price": float(quote.get("05. price", 0)),
            "change": float(quote.get("09. change", 0)),
            "change_percent": quote.get("10. change percent", "0%"),
            "volume": int(quote.get("06. volume", 0)),
            "timestamp": int(time.time()),
            "source": "alpha_vantage",
        }

    def _fetch_from_finnhub(self, symbol: str) -> MarketQuote:
        """Fetch data from Finnhub API."""
        url = "https://finnhub.io/api/v1/quote"
        params = {"symbol": symbol, "token": self.finnhub_key}

        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # Validate response has required fields
        required_fields = ["c", "d", "dp", "h", "l", "o", "pc", "t"]
        if not all(field in data for field in required_fields):
            raise ValueError("Invalid response from Finnhub")

        return {
            "symbol": symbol,
            "price": float(data["c"]),  # Current price
            "change": float(data["d"]),  # Change
            "change_percent": f"{float(data['dp']):.2f}%",  # Change percent
            "volume": 0,  # Finnhub doesn't provide volume in quote endpoint
            "timestamp": int(time.time()),
            "source": "finnhub",
        }

    def _fetch_from_polygon(self, symbol: str) -> MarketQuote:
        """Fetch data from Polygon API."""
        url = f"https://api.polygon.io/v2/aggs/ticker/{symbol}/prev"
        params = {"apiKey": self.polygon_key}

        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        if data.get("status") != "OK" or not data.get("results"):
            raise ValueError("Invalid response from Polygon")

        result = data["results"][0]
        return {
            "symbol": symbol,
            "price": float(result["c"]),  # Close price
            "change": 0.0,  # Polygon doesn't provide change in prev endpoint
            "change_percent": "0.00%",  # Would need previous close to calculate
            "volume": int(result["v"]),  # Volume
            "timestamp": int(time.time()),
            "source": "polygon",
        }


def create_market_data_provider(
    config: Optional[MarketDataConfig] = None,
) -> MarketDataService:
    """Factory function to create a market data provider."""
    return MarketDataService(config)
